bin/pytorch_inference/CBufferedIStreamAdapter.cc

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <core/CLogger.h>
+
+#include "CBufferedIStreamAdapter.h"
+
+// For ntohl
+#ifdef Windows
+#include <WinSock2.h>
+#else
+#include <netinet/in.h>
+#endif
+
+namespace ml {
+namespace torch {
+
+CBufferedIStreamAdapter::CBufferedIStreamAdapter(core::CNamedPipeFactory::TIStreamP inputStream) {
+    if (parseSizeFromStream(m_Size, inputStream) == false) {
+        LOG_ERROR(<< "Failed to read model size");
+    }
+
+    LOG_DEBUG(<< "Loading model of size: " << m_Size);
+
+    m_Buffer = std::make_unique<char[]>(m_Size);
+    inputStream->read(m_Buffer.get(), m_Size);
```

Comment:
When I first tested this I catted the model and piped it into the app on stdin and that was very slow taking 50 seconds to read a 1.2GB model. Increasing the underlying buffer size on stdin did not improve that time any.

When I moved the Python script which streams the model in 8KB chunks the loading time fell from 50 seconds to 1 second. 

I don't know why piping the model to stdin was so slow, perhaps because of this single large read??

Reply:
It’s libc++ that’s to blame - see https://bugs.llvm.org/show_bug.cgi?id=21192 - so it’s macOS-only.

Named pipe/disk file/stdin would be much closer in performance on Linux.

I am sure there is more historical research and conversations about this problem in old Prelert repos/Jira/GitLab.

But since we won’t be using stdin in production don’t spend any more time on this.